pkg/metrics/metrics.go

```diff
@@ -0,0 +1,145 @@
+	"github.com/jetstack/cert-manager/pkg/util/kube"
+)
+
+const (
+	// Namespace is the namespace for cert-manager metric names
+	namespace                              = "certmanager"
+	prometheusMetricsServerAddress         = "0.0.0.0:9402"
+	prometheusMetricsServerShutdownTimeout = 5 * time.Second
+	prometheusMetricsServerReadTimeout     = 8 * time.Second
+	prometheusMetricsServerWriteTimeout    = 8 * time.Second
+	prometheusMetricsServerMaxHeaderBytes  = 1 << 20 // 1 MiB
+
+	operationIssue = "issue"
+	operationRenew = "renew"
+)
+
+// Default set of metrics
+var Default = New()
+
+type Metrics struct {
+	http.Server
+
+	// TODO (@dippynark): switch this to use an interface to make it testable
+	registry                     *prometheus.Registry
+	CertificateExpiryTimeSeconds *prometheus.GaugeVec
+}
+
+func New() *Metrics {
+
+	router := mux.NewRouter()
+
+	// Create server and register prometheus metrics handler
+	s := &Metrics{
+		Server: http.Server{
+			Addr:           prometheusMetricsServerAddress,
+			ReadTimeout:    prometheusMetricsServerReadTimeout,
+			WriteTimeout:   prometheusMetricsServerWriteTimeout,
+			MaxHeaderBytes: prometheusMetricsServerMaxHeaderBytes,
+			Handler:        router,
+		},
+		registry: prometheus.NewRegistry(),
+		CertificateExpiryTimeSeconds: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: namespace,
+				Name:      "certificate_expiration_seconds",
+				Help:      "The date after which the certificate expires. Expressed as a Unix Epoch Time.",
+			},
+			[]string{"name", "namespace"},
+		),
+	}
+
+	s.registry.MustRegister(s.CertificateExpiryTimeSeconds)
+
+	router.Handle("/metrics", promhttp.HandlerFor(s.registry, promhttp.HandlerOpts{}))
+
+	return s
+}
+
+func (m *Metrics) waitShutdown(stopCh <-chan struct{}) {
+	<-stopCh
+	glog.Info("Stopping Prometheus metrics server...")
+
+	ctx, cancel := context.WithTimeout(context.Background(), prometheusMetricsServerShutdownTimeout)
+	defer cancel()
+
+	if err := m.Shutdown(ctx); err != nil {
+		glog.Errorf("Prometheus metrics server shutdown error: %v", err)
+		return
+	}
+
+	glog.Info("Prometheus metrics server gracefully stopped")
+}
+
+func (m *Metrics) Start(stopCh <-chan struct{}) {
+	go func() {
+
+		glog.Infof("Listening on http://%s", m.Addr)
+		if err := m.ListenAndServe(); err != nil {
+			glog.Errorf("Error running prometheus metrics server: %s", err.Error())
+			return
+		}
+
+		glog.Infof("Prometheus metrics server exited")
+
+	}()
+
+	m.waitShutdown(stopCh)
+}
+
+// UpdateCertificateExpiry updates the expiry time of a certificate
+func (m *Metrics) UpdateCertificateExpiry(crt *v1alpha1.Certificate, secretLister corelisters.SecretLister) {
+
+	// grab existing certificate and validate private key
+	cert, err := kube.SecretTLSCert(secretLister, crt.Namespace, crt.Spec.SecretName)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("[%s/%s] Error getting certificate '%s': %s", crt.Namespace, crt.Name, crt.Spec.SecretName, err.Error()))
```

Comment:
Check for `k8s.io/apimachinery/pkg/api/errors.IsNotFound` and `IsInvalidData` here else we'll log confuse non-error messages to users.